`./pants --changed-since=<ref> filter` crashes when a target points at a file with `@` in its name

**Describe the bug**

##### Background
(Minimized/generalized)

Our Django apps typically contain a `static/` folder containing images etc for use in templates. The folder is structured like:
```
static/
    BUILD.pants
    ...
    images/
        ....
        img1_80x80@2x.png
```

Since there are many sub-folders under `static/` and none of them are code, we've been lazy and putting:
```
resources(sources=["**/*"])
```
In `static/BUILD.pants`. Our templates / views can then depend on `./static` and pull in everything.

##### Problem

In this setup, if I run `./pants --changed-since=master filter`, the process crashes with:
```
17:37:34.92 [ERROR] Failed to parse Address `src/django_apps/core/static/images/site/cancer_80x80@2x.png:../../static`: error at 1:54: expected one or more key=value pairs to follow a `@`.
Traceback (most recent call last):
  File "/Users/dan.moran/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.3nnXTF/install/lib/python3.9/site-packages/pants/bin/daemon_pants_runner.py", line 131, in single_daemonized_run
    runner = LocalPantsRunner.create(
  File "/Users/dan.moran/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.3nnXTF/install/lib/python3.9/site-packages/pants/bin/local_pants_runner.py", line 156, in create
    specs = calculate_specs(
  File "/Users/dan.moran/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.3nnXTF/install/lib/python3.9/site-packages/pants/init/specs_calculator.py", line 71, in calculate_specs
    address_input = AddressInput.parse(address.spec)
  File "/Users/dan.moran/.cache/pants/setup/bootstrap-Darwin-x86_64/pants.3nnXTF/install/lib/python3.9/site-packages/pants/build_graph/address.py", line 141, in parse
    ) = native_engine.address_parse(spec)
native_engine.AddressParseException: Failed to parse Address `src/django_apps/core/static/images/img1_80x80@2x.png:../static`: error at 1:54: expected one or more key=value pairs to follow a `@`.
```

**Pants version**
v2.10.0

**OS**
Both MacOS and Linux

**Additional info**
* `./pants filter ::` doesn't crash in the same way
* `./pants filter --target-type=resource ::` doesn't crash, _and_ it successfully includes `static/images/img1_80x80@2x.png` in the output. So the `@` symbol doesn't always make things crash.

--- a/src/python/pants/build_graph/address_test.py
+++ b/src/python/pants/build_graph/address_test.py
@@ -75,6 +75,10 @@ def test_address_input_parse_spec() -> None:
     assert_parsed(
         "a#gen@k1=v1", generated_component="gen", path_component="a", parameters={"k1": "v1"}
     )
+    assert_parsed("a@t", path_component="a@t")
+    assert_parsed("a@=", path_component="a@=")
+    assert_parsed("a@t,y", path_component="a@t,y")
+    assert_parsed("a@2.png:../t", path_component="a@2.png", target_component="../t")
 
     # Absolute spec
     assert_parsed("//a/b/c", path_component="a/b/c")
@@ -112,9 +116,6 @@ def test_address_input_parse_bad_path_component(spec: str) -> None:
 @pytest.mark.parametrize(
     "spec,expected",
     [
-        ("a@t", "one or more key=value pairs"),
-        ("a@=", "one or more key=value pairs"),
-        ("a@t,y", "one or more key=value pairs"),
         ("a@t=,y", "one or more key=value pairs"),
         ("a#", "non-empty generated target name"),
     ],

--- a/src/rust/engine/address/src/lib.rs
+++ b/src/rust/engine/address/src/lib.rs
@@ -43,7 +43,8 @@ pub struct SpecInput<'a> {
 
 peg::parser! {
     grammar parsers() for str {
-        rule path() -> &'input str = s:$([^':' | '@' | '#']*) { s }
+        rule path() -> &'input str =
+            s:$(([^':' | '@' | '#'] / ("@" !parameter()))*) { s }
 
         rule target_name() -> &'input str
             = quiet!{ s:$([^'#' | '@' | ':']+) { s } }

